### PR TITLE
Refactor order filters

### PR DESCRIPTION
**Pull request by fowczarek**
I want to merge this change because refactoring orders filters:
- [x] refactor filtering by customer
- [x] refactor order search
- [x] drop all usage of `filter_by_query_param` 
- [x] Add db indexing

This PR will be merged to the feature branch, to merge migration it with #7301 

# Impact

* [x] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [x] Privileged queries and mutations are guarded by proper permission checks
* [x] Database queries are optimized and the number of queries is constant
* [x] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-3114 Drop usage of filter_by_query_param in order filters](https://app.clickup.com/t/2549495/SALEOR-3114)

File: saleor/account/models.py [hunk 1]
```diff
@@ -160,12 +161,17 @@ class User(PermissionsMixin, ModelWithMetadata, AbstractBaseUser):
...
         ordering = ("email",)
         permissions = (
             (AccountPermissions.MANAGE_USERS.codename, "Manage customers."),
             (AccountPermissions.MANAGE_STAFF.codename, "Manage staff."),
         )
+        indexes = [
+            # Orders searching index
+            GinIndex(fields=["email", "first_name", "last_name"]),
+        ]
+        indexes.extend(ModelWithMetadata.Meta.indexes)
```
Comment (tomaszszymanski129) [inline saleor/account/models.py:174]: we usually keep stuff in meta as single objects, maybe that would be more consistent?
```
indexes = [
    *ModelWithMetadata.Meta.indexes,
    # Orders searching index
    GinIndex(fields=["email", "first_name", "last_name"]),
]
```

File: saleor/graphql/order/filters.py [hunk 2]
```diff
@@ -26,6 +28,13 @@ def get_payment_id_from_query(value):
         return None
 
 
+def get_order_id_from_query(value):
+    if value.startswith("#"):
```
Comment (tomaszszymanski129) [inline saleor/graphql/order/filters.py:32]: little comment why this can start with `#`?

File: saleor/graphql/order/filters.py [hunk 3]
```diff
@@ -26,6 +28,13 @@ def get_payment_id_from_query(value):
...
+        value = value[1:]
+    if value.isnumeric():
+        return value
```
Comment (tomaszszymanski129) [inline saleor/graphql/order/filters.py:35]: Maybe `return value if value.isnumeric() else None` ? Personally I prefer to see that None can also be returned explicitly but maybe that's just my habit.

File: saleor/graphql/order/filters.py [hunk 4]
```diff
@@ -66,20 +74,29 @@ def filter_created_range(qs, _, value):
...
-        "payments__transactions__searchable_key",
-    ]
     payment_id = get_payment_id_from_query(value)
     if payment_id:
         return filter_order_by_payment(qs, payment_id)
 
-    qs = filter_by_query_param(qs, value, order_fields)
+    order_id = get_order_id_from_query(value)
+    if order_id:
+        return qs.filter(pk=order_id)
```
Comment (tomaszszymanski129) [inline saleor/graphql/order/filters.py:83]:
```
if order_id := get_order_id_from_query(value):
    return qs.filter(pk=order_id)
```

File: saleor/graphql/order/filters.py [hunk 5]
```diff
@@ -66,20 +74,29 @@ def filter_created_range(qs, _, value):
...
+
+    transactions = Transaction.objects.filter(searchable_key=value).values("id")
+    payments = Payment.objects.filter(
+        Exists(transactions.filter(payment_id=OuterRef("id")))
+    )
```
Comment (tomaszszymanski129) [inline saleor/graphql/order/filters.py:88]: shouldn't this end with `.values`?

File: saleor/graphql/order/filters.py [hunk 6]
```diff
@@ -66,20 +74,29 @@ def filter_created_range(qs, _, value):
...
+    ).values("id")
+    qs = qs.filter(
+        Q(user_email__trigram_similar=value)
+        | Q(Exists(discounts.filter(order_id=OuterRef("id"))))
+        | Q(user__email__trigram_similar=value)
+        | Q(user__first_name__trigram_similar=value)
+        | Q(user__last_name__trigram_similar=value)
+        | Q(Exists(payments.filter(order_id=OuterRef("id"))))
+    )
     return qs
```
Comment (tomaszszymanski129) [inline saleor/graphql/order/filters.py:100]: `return qs.filter(...)`

File: saleor/order/models.py [hunk 7]
```diff
@@ -224,9 +225,11 @@ class Order(ModelWithMetadata):
     redirect_url = models.URLField(blank=True, null=True)
     objects = OrderQueryset.as_manager()
 
-    class Meta(ModelWithMetadata.Meta):
+    class Meta:
         ordering = ("-pk",)
         permissions = ((OrderPermissions.MANAGE_ORDERS.codename, "Manage orders."),)
+        indexes = [GinIndex(fields=["user_email"])]
+        indexes.extend(ModelWithMetadata.Meta.indexes)
```
Comment (tomaszszymanski129) [inline saleor/order/models.py:232]: above I suggested approach without using extend, also apply it here if you think it's better.

**Review by tomaszszymanski129 — CHANGES_REQUESTED** [review]

**Review by korycins — APPROVED** [review]

File: saleor/graphql/order/filters.py (hunk 2 above)
Comment (fowczarek) [reply saleor/graphql/order/filters.py:32]: In the dashboard, we display Order number with # as the prefix.
image

**Review by tomaszszymanski129 — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Review by patrys — APPROVED** [review]

**Pull request merged**